### PR TITLE
Add extra values to the ExecutionContext class

### PR DESCRIPTION
**Pull request by jkimbo**
## Description

This PR does a couple of things:

* Add extra values to the `ExecutionContext` type so that extensions can make use of it
* Pass the execution context instance to the extension so that it can be accessed inside the extension hooks (also remove it from being passed in as an argument to the hooks)
* Moves the creation of the `ExecutionContext` type to the schema (instead of in the execute functions)
* Replaces the `ExecutionContext` type creation in the views to instead use a helper function `parse_request_data` to extra the relevant request data from the input json
* Pass the execution context to the `process_errors` function because it might be useful

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/),so make sure to pick the appropriate type. If in doubt feel free to ask :)

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/),so make sure to pick the appropriate type. If in doubt feel free to ask :)

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/),so make sure to pick the appropriate type. If in doubt feel free to ask :)

File: RELEASE.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+@dataclasses.dataclass
+class ExecutionContext:
+    query: str
+    context: Any = None
+    variables: Optional[Dict[str, Any]] = None
+    operation_name: Optional[str] = None
+
+    graphql_document: Optional[GraphQLDocumentNode] = None
+    errors: Optional[List[GraphQLError]] = None
+    result: Optional[GraphQLExecutionResult] = None
```
Comment (patrick91) [inline RELEASE.md:22]: we should probably abstract these in future 😊

File: RELEASE.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+and can be accessed in any of the extension hooks:
+
+```python
+from strawberry.extensions import Extension
+
+class MyExtension(Extension):
+    def on_request_end(self):
+        result = self.execution_context.result
+        # Do something with the result
```
Comment (patrick91) [inline RELEASE.md:33]: 👍

File: strawberry/asgi/__init__.py [hunk 3]
```diff
@@ -263,20 +264,18 @@ async def get_http_response(
             )
 
         try:
-            query = data["query"]
-            variables = data.get("variables")
-            operation_name = data.get("operationName")
-        except KeyError:
+            request_data = parse_request_data(data)
```
Comment (patrick91) [inline strawberry/asgi/__init__.py:267]: I like this 😊

File: strawberry/http.py [hunk 4]
```diff
@@ -19,3 +20,22 @@ def process_result(result: ExecutionResult) -> GraphQLHTTPResponse:
         data["errors"] = [format_graphql_error(err) for err in result.errors]
 
     return data
+
+
+class GraphQLRequestData(TypedDict, total=False):
+    query: str
+    variables: Optional[Dict[str, Any]]
+    operation_name: Optional[str]
```
Comment (patrick91) [inline strawberry/http.py:28]: why a dict and not a dataclass? 😊

File: tests/schema/extensions/test_extensions.py [hunk 5]
```diff
@@ -96,3 +96,72 @@ async def person(self) -> Person:
...
+
+
+def test_extension_access_to_parsed_document():
+    query_name = ""
+
+    class MyExtension(Extension):
+        def on_parsing_end(self):
+            nonlocal query_name
+            query_defintion = self.execution_context.graphql_document.definitions[0]
+            query_name = query_defintion.name.value
```
Comment (patrick91) [inline tests/schema/extensions/test_extensions.py:108]:
```suggestion
            query_definition = self.execution_context.graphql_document.definitions[0]
            query_name = query_definition.name.value
```

**Review by patrick91 — APPROVED** [review]
Nice one! I've added a couple of minor comments, nothing that prevents this to get merged 😊

File: strawberry/http.py (hunk 4 above)
Comment (jkimbo) [reply strawberry/http.py:28]: No particular reason. I saw another TypedDict in this file so I just copied that. I'll change it to a dataclass.

File: tests/schema/extensions/test_extensions.py (hunk 5 above)
Comment (jkimbo) [reply tests/schema/extensions/test_extensions.py]: Good catch!

File: strawberry/http.py (hunk 4 above)
Comment (patrick91) [reply strawberry/http.py:28]: ok!

**Pull request merged**